doc/configurator.rst

```diff
@@ -0,0 +1,113 @@
+************
+Configurator
+************
+
+Configurator is a small library designed to query features available on the
+system, in order to generate configuration for dune builds. Such generated
+configuration is usually in the form of command line flags, generated headers,
+stubs, but there are no limitations on this.
+
+Configurator allows you to query for the following features:
+
+* Variables defined in ``ocamlc - config``
+
+* pkg-config_ flags for packages
+
+* Test features by compiling C code
+
+* Extract compile time information such as ``#define`` variables.
+
+Configurator is designed to be cross compilation friendly and avoids running any
+compiled code to extract any of the information above.
+
+Configurator started as an `independent library
+<https://github.com/janestreet/configurator>`__, but now lives in dune. You do
+not need to install anything to use configurator.
+
+Usage
+=====
+
+We'll describe configurator with a simple example. Everything else can be easily
+learned by studying configurator's API.
+
+To use configurator, we write an executable that will query the system using
+configurator's API and output a set of targets reflecting the results. For
+example:
+
+.. code-block:: ocaml
+
+  module C = Configurator.V1
+
+  let clock_gettime_code = {|
+  #include <time.h>
+
+  int main()
+  {
+    struct timespec ts;
+    clock_gettime(CLOCK_REALTIME, &ts);
+    return 0;
+  }
+  |}
+
+  let () =
+    C.main ~name:"foo" (fun c ->
+      let has_clock_gettime = C.c_test c clock_gettime_code ~link_flags:["-lrt"] in
+
+      C.C_define.gen_header_file c ~fname:"config.h"
+        [ "HAS_CKOCK_GETTIME", Switch has_ckock_gettime ]);
+
+Usually, the module above would be named ``discover.ml``. The next step is to
+invoke it as an executable and tell dune about the targets that it produces:
+
+.. code-block:: lisp
+
+  (executable
+   ((name discover)
+    (libraries (dune.configurator))))
+
+  (rule
+   ((targets (config.h))
+    (action (run ./discover.exe))))
+
+Another common pattern is to produce a flags file with configurator and then use
```

Comment:
This part is not obvious from the API and people will refrain to use `Usexp` since it is labeled _internal_.  In many packages, `Stdio` is used but since Dune already comes bundled with `Usexp`, shouldn't we use it — or provide something minimal in Configurator based on Usexp?

Reply:
I agree that we should integrate `Usexp` better. I don't think @diml deems `Usexp` as ready for public consumption so once it's made available, we can recommend it.

Reply:
I think usexp is fine as it is actually, I was more thinking of the sexp module in src. For configurator, it probably makes sense to add a function `val write_flags : string -> string list -> unit` to dump a list of strings into a file in S-expression syntax.